ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java

```diff
@@ -45,45 +46,45 @@
   private volatile boolean shutdown;
   private final WatchService watchService;
   private final Path file;
+  private final Path dir;
   private final Callback callback;
+  private final WatchKey key;
 
   @SuppressFBWarnings(
       value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE",
       justification = "Null check on file.getParent() is present"
   )
   public FileWatcher(final Path file, final Callback callback) throws IOException {
     this.file = Objects.requireNonNull(file);
-    Objects.requireNonNull(file.getParent(), "Watch location must have parent");
+    this.dir = Objects.requireNonNull(file.getParent(), "Watch location must have parent");
     this.watchService = FileSystems.getDefault().newWatchService();
+    this.callback = Objects.requireNonNull(callback);
+
     // Listen to both CREATE and MODIFY to reload, which handles delete then create.
-    file.getParent().register(watchService,
+    this.key = dir.register(watchService,
         StandardWatchEventKinds.ENTRY_CREATE,
         StandardWatchEventKinds.ENTRY_MODIFY);
-    this.callback = Objects.requireNonNull(callback);
   }
 
   /**
    * Closes the file watcher
    */
   public void shutdown() {
+    log.info("Stopping file watcher from watching for changes: " + file);
     shutdown = true;
   }
 
   @Override
   public void run() {
     log.info("Starting file watcher to watch for changes: " + file);
     try {
-      while (!shutdown) {
+      while (!shutdown && key.isValid()) {
         try {
           handleNextWatchNotification();
-        } catch (InterruptedException e) {
-          throw e;
         } catch (Exception e) {
```

Comment:
why did we change this to swallow interrupted exceptions? seems like an atni-pattern

Reply:
Because the code doesn't interrupt the thread.

Reply:
I still don't think that we should be swallowing the interrupted exception - I don't see the benefit (we lose nothing by keeping it, we might lose something by removing it). Just because the code today doesn't intentionally interrupt this thread, it doesn't mean that future code won't or that there is some code we didn't audit that does.

Reply:
It's generally considered bad practice to us interrupts to control thread lifecycles.  Given we're not, if someone wants to shutdown this thread they should call `shutdown` on the watcher and the code will do the right thing.  If they try interrupting the thread then the code will log a `unexpected exception` error and continue - this to me is the _right_ behaviour: interrupt exceptions aren't expected and shouldn't be used to control the lifecycle of this thread.

Reply:
I didn't mean to suggest that we should shutdown the thread on interrupt but at we should catch it and reset the thread's interrupt flag to make sure we don't loose this information. I'll defer to Brian Goetz's article as he makes the point way better than I can: https://www.ibm.com/developerworks/library/j-jtp05236/index.html

The relevant part:
> If you catch InterruptedException but cannot rethrow it, you should preserve evidence that the interruption occurred so that code higher up on the call stack can learn of the interruption and respond to it if it wants to. This task is accomplished by calling interrupt() to "reinterrupt" the current thread, as shown in Listing 3. At the very least, whenever you catch InterruptedException and don't rethrow it, reinterrupt the current thread before returning. 
>
> The worst thing you can do with InterruptedException is swallow it -- catch it and neither rethrow it nor reassert the thread's interrupted status. The standard approach to dealing with an exception you didn't plan for -- catch it and log it -- also counts as swallowing the interruption because code higher up on the call stack won't be able to find out about it. (Logging InterruptedException is also just silly because by the time a human reads the log, it is too late to do anything about it.)